File: bfs/0987_hard/vertical_traversal.cpp
#include <vector>
#include <list>
#include <map>
#include <unordered_map>
#include <queue>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

vector<vector<int>> verticalTraversal(TreeNode *root) // LeetCode Q.987.
{
    map<int, vector<int>> cols2nums;

    queue<pair<TreeNode *, int>> queue; // Format: {node, col}.
    queue.push({root, 0});

    unordered_map<int, vector<int>> current_row_values; // Keys: cols.
    list<pair<TreeNode *, int>> next_row_nodes;         // Format: {node, col}.

    while (!queue.empty())
    {
        TreeNode *node = queue.front().first;
        int col = queue.front().second;

        queue.pop();
        if (node->left != nullptr)
        {
            next_row_nodes.push_back({node->left, col - 1});
        }
        if (node->right != nullptr)
        {
            next_row_nodes.push_back({node->right, col + 1});
        }

        if (cols2nums.find(col) == cols2nums.end())
        {
            cols2nums[col] = {};
        }
        if (current_row_values.find(col) == current_row_values.end())
        {
            current_row_values[col] = {};
        }
        current_row_values[col].push_back(node->val);
        sort(current_row_values[col].begin(), current_row_values[col].end());

        if (queue.empty())
        {
            while (!next_row_nodes.empty())
            {
                pair<TreeNode *, int> next_row_node = next_row_nodes.front();
                queue.push(next_row_node);
                next_row_nodes.pop_front();
            }
            for (const auto &pair : current_row_values)
            {
                for (auto value : pair.second)
                {
                    cols2nums[pair.first].push_back(value);
                }
            }
            current_row_values.clear(); // Clear for the next row.
        }
    }

    vector<vector<int>> vectical_traversal;
    for (const auto &pair : cols2nums)
    {
        vectical_traversal.push_back(pair.second);
    }
    return vectical_traversal;
}

File: bfs/0987_hard/vertical_traversal.py

class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val = val
        self.left = left
        self.right = right


def find_vertical_traversal(root: TreeNode | None) -> list[list[int]]:  # LeetCode Q.987.
    queue = []  # Format: (node, col).
    if root is not None:
        queue.append((root, 0))

    cols2nums: dict[int, list[int]] = dict()
    next_level_nodes = []  # Format: (node, col).
    current_row_values: dict[int, list[int]] = dict()  # Keys: cols.

    while queue:
        node, col = queue.pop(0)
        if node.left:
            next_level_nodes.append((node.left, col - 1))
        if node.right:
            next_level_nodes.append((node.right, col + 1))

        if col not in cols2nums.keys():
            cols2nums[col] = []

        if col not in current_row_values.keys():
            current_row_values[col] = []
        current_row_values[col].append(node.val)
        current_row_values[col].sort()

        if not queue:  # Current level BFS is done.
            queue.extend(next_level_nodes)  # Put next level into queue.
            next_level_nodes.clear()

            for col, values in current_row_values.items():
                cols2nums[col].extend(values)
            current_row_values.clear()

    cols2nums = dict(sorted(cols2nums.items()))
    vertical_traversal = []
    for col, nums in cols2nums.items():
        vertical_traversal.append(nums)
    return vertical_traversal
